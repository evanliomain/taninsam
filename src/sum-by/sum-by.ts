/**
 * @module array=>number
 */
import { ChainFunction, Iteratee, Links } from '../@types';
/**
 * This method sum all element of the chained array,
 * invoking for each element in array,
 * the iteratee callback to get the value to be summed.
 * @param iteratee The iteratee invoked per element to get the value
 * to sum given an element of the array
 * @return the function to apply on the array
 * to sum its elements and return the sum
 * @example With an array of number
 * ```
 * sumBy<number>(x => x)([1, 2, 3, 4]) // 10
 * ```
 * @example With an array of object
 * ```
 * sumBy<{ a: number }>(x => x.a)(
 *  [{ a: 1 }, { a: 2 }, { a: 3 }, { a: 4 }]
 * ) // 10
 * ```
 * @example Using the chain
 * ```
 * chain([{ a: 1 }, { a: 2 }, { a: 3 }, { a: 4 }])
 *   .chain(sumBy<{ a: number }>(x => x.a))
 *   .value() // 10
 * ```
 */
export function sumBy<T>(
  iteratee: Iteratee<T, number>,
): ChainFunction<ReadonlyArray<T>, number> {
  return (array: ReadonlyArray<T>, links?: Links) =>
    array.reduce(
      (sum, data, index: number, originalArray: ReadonlyArray<T>) =>
        sum + iteratee(data, index, originalArray, links),
      0,
    );
}
