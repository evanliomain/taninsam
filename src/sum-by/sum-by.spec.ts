import { sumBy } from './sum-by';

describe('sumBy function', () => {
  const input = [1, 2, 3, 4, 5];

  test('is a pure function', () => {
    sumBy<number>(x => x)(input);
    expect(input).toMatchSnapshot();
  });

  test('[{ a:1 }, { a:2 }, { a:3 }, { a:4 }] |> sumBy === 10', () => {
    expect(
      sumBy<{ a: number }>(x => x.a)([{ a: 1 }, { a: 2 }, { a: 3 }, { a: 4 }])
    ).toMatchSnapshot();
  });

  test('call the iteree callback with correct inputs', () => {
    const cb = jest.fn((x: number) => x);
    const link = { fake: 'fake' };
    sumBy<number>(cb)(input, link);
    input.forEach((i, index) => {
      expect(cb.mock.calls[index][0]).toBe(i);
      expect(cb.mock.calls[index][1]).toBe(index);
      expect(cb.mock.calls[index][2]).toEqual(input);
      expect(cb.mock.calls[index][3]).toBe(link);
    });
  });
});
