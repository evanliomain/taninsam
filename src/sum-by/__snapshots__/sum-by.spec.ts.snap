// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sumBy function > [{ a:1 }, { a:2 }, { a:3 }, { a:4 }] |> sumBy === 10 1`] = `10`;

exports[`sumBy function > is a pure function 1`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;
